src/app/shared/spaces.service.ts

```diff
@@ -1,72 +1,80 @@
 import { Injectable } from '@angular/core';
-
-import { LocalStorageService } from 'angular-2-local-storage';
 import { Broadcaster } from 'ngx-base';
 import { Contexts, Space, Spaces, SpaceService } from 'ngx-fabric8-wit';
-import { Observable, Subject } from 'rxjs';
-import { ConnectableObservable } from 'rxjs/observable/ConnectableObservable';
-import { ReplaySubject } from 'rxjs/ReplaySubject';
-
+import { Observable, Subscription } from 'rxjs';
 import { ExtProfile, ProfileService } from '../profile/profile.service';
 
 @Injectable()
 export class SpacesService implements Spaces {
 
   static readonly RECENT_SPACE_LENGTH = 8;
-  private static readonly RECENT_SPACE_KEY = 'recentSpaceIds';
 
-  addRecent: Subject<Space> = new Subject<Space>();
   private _current: Observable<Space>;
-  private _recent: ConnectableObservable<Space[]>;
+  private _recent: Observable<Space[]>;
 
   constructor(
     private contexts: Contexts,
-    private localStorage: LocalStorageService,
     private spaceService: SpaceService,
     private broadcaster: Broadcaster,
     private profileService: ProfileService
   ) {
+    this.initCurrent();
+    this.initRecent();
+  }
+
+  private initCurrent(): void {
     this._current = Observable.merge(
-      contexts.current
-        .map(val => val.space),
-      this.broadcaster.on<Space>('spaceUpdated'));
-    // Create the recent context list
+      this.contexts.current.map(val => val.space),
+      this.broadcaster.on<Space>('spaceUpdated')
+    );
+  }
+
+  private initRecent(): void {
+    // load existing recent spaces from profile.store.recentSpaces
+    this._recent = this.loadRecent();
 
-    this._recent = Observable.merge(
-      this.broadcaster.on<Space>('spaceChanged'),
-      this.addRecent)
-      // Map from the context being added to an array of recent contexts
-      // The scan operator allows us to access the list of recent contexts and add ours
-      .scan((recent, s) => {
-        // First, check if this context is already in the list
-        // If it is, remove it, so we don't get duplicates
-        for (let i = recent.length - 1; i >= 0; i--) {
-          if (recent[i].id === s.id) {
-            recent.splice(i, 1);
+    // update _recent with newly seen space when spaceChanged is emitted
+    this.broadcaster.on<Space>('spaceChanged')
+      .combineLatest(this._recent)
```

Comment:
This usage of combineLatest isn't necessary as the array is accessible already. The broadcast event will emit the new space, and can be subscribed directly to receive it, then manipulating the array with the new space.

Reply:
Similar to the `spaceDeleted` broadcast comment below, this `combineLatest` looks like it should really be a mapping, with the subscriber logic moved up into the mapping and the subscriber responsible just for saving.